sapi/cli/php_cli_server.c

```diff
@@ -1572,50 +1575,95 @@ static int php_cli_server_client_read_request_on_fragment(php_http_parser *parse
 {
 	php_cli_server_client *client = parser->data;
-	char *value = pestrndup(at, length, 1);
-	if (!value) {
-		return 1;
-	}
-	{
-		/* strip off the colon */
-		zend_string *orig_header_name = zend_string_init(client->current_header_name, client->current_header_name_len, 1);
-		char *lc_header_name = zend_str_tolower_dup(client->current_header_name, client->current_header_name_len);
-		zend_hash_str_add_ptr(&client->request.headers, lc_header_name, client->current_header_name_len, value);
-		zend_hash_add_ptr(&client->request.headers_original_case, orig_header_name, value);
-		efree(lc_header_name);
-		zend_string_release(orig_header_name);
+	switch (client->last_header_element) {
+	case VALUE:
+		php_cli_server_client_save_header(client);
+	case NONE:
+		client->current_header_name = at;
+		client->current_header_name_len = length;
+		break;
+	case FIELD:
+		if (client->current_header_name_allocated) {
+			size_t new_length = client->current_header_name_len + length;
+			client->current_header_name = perealloc(client->current_header_name, new_length + 1, 1);
+			strncpy(client->current_header_name + client->current_header_name_len, at, length);
+			client->current_header_name_len = new_length;
+		} else {
+			size_t new_length = client->current_header_name_len + length;
+			char* field = pemalloc(new_length + 1, 1);
+			strncpy(field, client->current_header_name, client->current_header_name_len);
+			strncpy(field + client->current_header_name_len, at, length);
+			client->current_header_name = field;
+			client->current_header_name_len = new_length;
+			client->current_header_name_allocated = 1;
+		}
+		break;
 	}
 
-	if (client->current_header_name_allocated) {
-		pefree(client->current_header_name, 1);
-		client->current_header_name_allocated = 0;
+	client->last_header_element = FIELD;
+	return 0;
+}
+
+static int php_cli_server_client_read_request_on_header_value(php_http_parser *parser, const char *at, size_t length)
+{
+	php_cli_server_client *client = parser->data;
+	switch (client->last_header_element) {
+	case FIELD:
+		client->current_header_value = pestrndup(at, length, 1);
+		client->current_header_value_len = length;
+		break;
+	case VALUE:
+		{
+			size_t new_length = client->current_header_value_len + length;
+			client->current_header_value = perealloc(client->current_header_value, new_length + 1, 1);
+			strncpy(client->current_header_value + client->current_header_value_len, at, length);
+			client->current_header_value_len = new_length;
+		}
+		break;
+	case NONE:
+		// can't happen
+		assert(1);
+		break;
 	}
+	client->last_header_element = VALUE;
 	return 0;
 }
 
 static int php_cli_server_client_read_request_on_headers_complete(php_http_parser *parser)
 {
 	php_cli_server_client *client = parser->data;
-	if (client->current_header_name_allocated) {
-		pefree(client->current_header_name, 1);
-		client->current_header_name_allocated = 0;
+	switch (client->last_header_element) {
+	case NONE:
+		break;
+	case FIELD:
```

Comment:
Can the FIELD case occur here and if so, will it be handled correctly?

Reply:
Yes it can occur and it needs to be handled appropriately, fixed